src/message/message-queue.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+
+const messageParser = require('./message-parser')
+const C = require('../constants/constants')
+
+const STATE = {
+  IDLE: 0,
+  ASYNC: 1,
+  SYNC: 2,
+  WAITING: 3,
+}
+
+module.exports = class MessageQueue {
+
+  constructor(options, socket) {
+    this._processor = options.processor
+    this._permissionHandler = options.permissionHandler
+    this._logger = options.logger
+    this._socket = socket
+    this._messages = []
+    this._index = 0
+    this._state = STATE.IDLE
+
+    this._onMessage = this._onMessage.bind(this)
+    this._onResponse = this._onResponse.bind(this)
+  }
+
+  /**
+   * There will only ever be one consumer of forwarded messages. So rather than using
+   * events - and their performance overhead - the messageProcessor exposes
+   * this method that's expected to be overwritten.
+   *
+   * @param   {SocketWrapper} socketWrapper
+   * @param   {Object} message the parsed message
+   *
+   * @overwrite
+   *
+   * @returns {void}
+   */
+  onAuthenticatedMessage(socketWrapper, message) {
+  }
+
+  /**
+   * This method is the way the message queue accepts input. It receives arrays
+   * of parsed messages, iterates through them and issues permission requests for
+   * each individual message
+   *
+   * @todo The responses from the permissionHandler might arive in any arbitrary order - order them
+   * @todo Handle permission handler timeouts
+   *
+   * @param   {SocketWrapper} socketWrapper
+   * @param   {Object} message parsed message
+   *
+   * @returns {void}
+   */
+  process(rawMessage) {
+    messageParser.parse(rawMessage, this._onMessage)
```

Comment:
@ronag is there a reason this now does a callback? it seems pretty much identical as returning an array of messages.